libnetwork/datastore/datastore.go

```diff
@@ -346,10 +345,8 @@ func (ds *datastore) PutObjectAtomic(kvObject KVObject) error {
 		pair     *store.KVPair
 		err      error
 	)
-	if ds.sequential {
-		ds.Lock()
-		defer ds.Unlock()
-	}
+	ds.Lock()
+	defer ds.Unlock()
```

Comment:
https://github.com/moby/libnetwork/issues/1201#issuecomment-223185754
> The main purpose of serializing access was for local store where we saw data corruptions.

Sounds like there are some concurrency issues with datastore which the serialization was added to work around. I have a hunch that it involves cache coherency or an impedance mismatch with the libkv API not supporting multi-operation transactions. I hope this cleanup and refactoring work will get us closer to being able to remove the locking.